File: protocol/tools/compatibility/schema_checks/checks.py
"""What the schema has to say: the exact objects each envelope kind declares,
the fields that must be present, and the shapes the meta and error members
take. This is the part that fails when the protocol changes without its
checker being told."""

from __future__ import annotations

from typing import Any, Sequence

from ..contract import (
    CAMEL_FIELDS,
    ERROR_REQUIRED,
    KINDS,
    META_OPTIONAL,
    META_REQUIRED,
    PROTOCOL_ID,
    REPLAY_OPTIONAL,
    REPLAY_REQUIRED,
    REQUIRED_FIELDS,
    Document,
)
from ..documents import (
    _object_schema_for_kind,
    _object_schema_with_properties,
    _resolved_schema,
    _schema_with_property_const,
)

def _check_exact_object(
    schema: Any,
    *,
    label: str,
    required: Sequence[str],
    optional: Sequence[str] = (),
    errors: list[str],
) -> None:
    if not isinstance(schema, dict):
        errors.append(f"{label}: expected an object schema")
        return
    properties = schema.get("properties")
    actual_required = schema.get("required")
    if schema.get("type") != "object":
        errors.append(f"{label}: type must be 'object'")
    if schema.get("additionalProperties") is not False:
        errors.append(f"{label}: additionalProperties must be false")
    if not isinstance(properties, dict):
        errors.append(f"{label}: properties must be an object")
        return
    expected_properties = set(required) | set(optional)
    if set(properties) != expected_properties:
        errors.append(
            f"{label}: properties are {sorted(properties)}, expected {sorted(expected_properties)}"
        )
    if not isinstance(actual_required, list) or set(actual_required) != set(required):
        errors.append(f"{label}: required is {actual_required!r}, expected {list(required)!r}")


def _check_nonempty(schema: Any, label: str, errors: list[str]) -> None:
    if not isinstance(schema, dict) or schema.get("type") != "string" or schema.get("minLength") != 1:
        errors.append(f"{label}: must be a string schema with minLength 1")

def _check_arbitrary(schema: Any, label: str, errors: list[str]) -> None:
    if schema != {}:
        errors.append(f"{label}: must accept arbitrary JSON (expected an empty schema)")

def _check_schema_contract(
    documents: Sequence[Document], errors: list[str]
) -> dict[str, dict[str, Any]]:
    schemas: dict[str, dict[str, Any]] = {}
    for kind in KINDS:
        schema = _object_schema_for_kind(documents, kind)
        if schema is None:
            errors.append(f"schema: no '{kind}' envelope with properties.type.const == '{kind}'")
            continue
        schemas[kind] = schema
        optional = ("requestId",) if kind == "event" else (("id",) if kind == "error" else ())
        _check_exact_object(
            schema, label=f"schema {kind} envelope", required=REQUIRED_FIELDS[kind],
            optional=optional, errors=errors,
        )
        discriminator = schema.get("properties", {}).get("type")
        if not isinstance(discriminator, dict) or discriminator.get("const") != kind:
            errors.append(f"schema {kind} envelope: invalid type discriminator")

    request = schemas.get("request")
    if request:
        properties = request.get("properties", {})
        for name in ("id", "method"):
            _check_nonempty(_resolved_schema(documents, properties.get(name)), f"schema request.{name}", errors)
        _check_arbitrary(properties.get("params"), "schema request.params", errors)
        meta = _resolved_schema(documents, properties.get("meta"))
        _check_exact_object(
            meta, label="schema request.meta", required=META_REQUIRED,
            optional=META_OPTIONAL, errors=errors,
        )
        if isinstance(meta, dict):
            version = meta.get("properties", {}).get("protocolVersion")
            if not isinstance(version, dict) or version.get("const") != PROTOCOL_ID:
                errors.append(
                    f"schema request.meta.protocolVersion: const must be {PROTOCOL_ID!r}"
                )
            _check_nonempty(
                _resolved_schema(documents, meta.get("properties", {}).get("idempotencyKey")),
                "schema request.meta.idempotencyKey", errors,
            )

    response = schemas.get("response")
    if response:
        _check_nonempty(_resolved_schema(documents, response.get("properties", {}).get("id")), "schema response.id", errors)
        _check_arbitrary(response.get("properties", {}).get("result"), "schema response.result", errors)

    event = schemas.get("event")
    if event:
        properties = event.get("properties", {})
        for name in ("sessionId", "streamId", "cursor", "eventId", "kind"):
            _check_nonempty(_resolved_schema(documents, properties.get(name)), f"schema event.{name}", errors)
        sequence = properties.get("sequence")
        if not isinstance(sequence, dict) or sequence.get("type") != "integer" or sequence.get("minimum") != 0:
            errors.append("schema event.sequence: must be an integer with minimum 0")
        if "requestId" in properties:
            _check_nonempty(_resolved_schema(documents, properties.get("requestId")), "schema event.requestId", errors)
        _check_arbitrary(properties.get("payload"), "schema event.payload", errors)

    error = schemas.get("error")
    if error:
        properties = error.get("properties", {})
        if "id" in properties:
            _check_nonempty(_resolved_schema(documents, properties.get("id")), "schema error.id", errors)
        payload = _resolved_schema(documents, properties.get("error"))
        _check_exact_object(
            payload, label="schema error.error", required=ERROR_REQUIRED, errors=errors
        )
        if isinstance(payload, dict):
            nested = payload.get("properties", {})
            _check_arbitrary(nested.get("details"), "schema error.error.details", errors)
            retryable = nested.get("retryable")
            if not isinstance(retryable, dict) or retryable.get("type") != "boolean":
                errors.append("schema error.error.retryable: must be boolean")

    replay_marker = _schema_with_property_const(documents, "method", "session.replay")
    if replay_marker is None:
        errors.append("schema: no request specialization with method const 'session.replay'")
    params = _object_schema_with_properties(documents, set(REPLAY_REQUIRED) | set(REPLAY_OPTIONAL))
    if params is None:
        errors.append("schema: no session.replay params object with sessionId/cursor/limit")
    else:
        _check_exact_object(
            params, label="schema session.replay params", required=REPLAY_REQUIRED,
            optional=REPLAY_OPTIONAL, errors=errors,
        )
        _check_nonempty(
            _resolved_schema(documents, params.get("properties", {}).get("sessionId")),
            "schema replay.sessionId", errors,
        )
        limit = _resolved_schema(documents, params.get("properties", {}).get("limit"))
        if not isinstance(limit, dict) or (
            limit.get("type") != "integer" or limit.get("minimum", 0) < 0
        ):
            errors.append("schema replay.limit: must be a nonnegative integer")
    return schemas

File: tools/src/protocol/checks.rs
//! What the schema has to say: the exact objects each envelope kind declares,
//! the fields that must be present, and the shapes the meta, error and replay
//! members take. This is the part that fails when the protocol changes
//! without the contract in protocol/contract.json being told.

use super::documents::{object_with_properties, resolved, schema_with_property_const, Document};
use serde_json::{Map, Value};
use std::collections::{BTreeMap, BTreeSet};

pub(super) type Located<'a> = (usize, &'a Value);

fn names(value: &Value) -> Vec<String> {
    value
        .as_array()
        .map(|items| {
            items
                .iter()
                .filter_map(|item| item.as_str().map(str::to_string))
                .collect()
        })
        .unwrap_or_else(Vec::new)
}

/// An object schema with `additionalProperties: false`, whose `required`
/// list is exactly the contract's. When the contract lists `optional` names,
/// the declared properties must be exactly required plus optional; otherwise
/// they must at least include every required name.
fn exact_object(schema: &Value, label: &str, member: &Value, errors: &mut Vec<String>) {
    let Some(object) = schema.as_object() else {
        errors.push(format!("{label}: expected an object schema"));
        return;
    };
    if object.get("type").and_then(Value::as_str) != Some("object") {
        errors.push(format!("{label}: type must be 'object'"));
    }
    if object.get("additionalProperties") != Some(&Value::Bool(false)) {
        errors.push(format!("{label}: additionalProperties must be false"));
    }
    let Some(properties) = object.get("properties").and_then(Value::as_object) else {
        errors.push(format!("{label}: properties must be an object"));
        return;
    };
    let required = names(&member["required"])
        .into_iter()
        .collect::<BTreeSet<_>>();
    let declared = properties.keys().cloned().collect::<BTreeSet<_>>();
    if member["optional"].is_array() {
        let mut expected = required.clone();
        expected.extend(names(&member["optional"]));
        if declared != expected {
            errors.push(format!(
                "{label}: properties are {declared:?}, expected {expected:?}"
            ));
        }
    } else if !required.is_subset(&declared) {
        errors.push(format!(
            "{label}: properties are {declared:?}, which lack {:?}",
            required.difference(&declared).collect::<Vec<_>>()
        ));
    }
    let actual = object.get("required").map(names);
    if actual.map(|list| list.into_iter().collect::<BTreeSet<_>>()) != Some(required.clone()) {
        errors.push(format!(
            "{label}: required is {:?}, expected {required:?}",
            object.get("required")
        ));
    }
}

fn property<'a>(schema: &'a Value, name: &str) -> &'a Value {
    &schema["properties"][name]
}

/// The per-field shapes the contract names for one object schema.
fn field_shapes(
    root: &Value,
    schema: &Value,
    label: &str,
    member: &Value,
    errors: &mut Vec<String>,
) {
    for name in names(&member["nonEmpty"]) {
        let field = property(schema, &name);
        if field.is_null() && !names(&member["required"]).contains(&name) {
            continue;
        }
        let field = resolved(root, field);
        if field["type"] != "string" || field["minLength"] != 1 {
            errors.push(format!(
                "{label}.{name}: must be a string schema with minLength 1"
            ));
        }
    }
    for name in names(&member["arbitrary"]) {
        if property(schema, &name) != &Value::Object(Map::new()) {
            errors.push(format!(
                "{label}.{name}: must accept arbitrary JSON (expected an empty schema)"
            ));
        }
    }
    for name in names(&member["nonNegativeIntegers"]) {
        let field = resolved(root, property(schema, &name));
        if field["type"] != "integer"
            || !field["minimum"]
                .as_i64()
                .is_some_and(|minimum| minimum >= 0)
        {
            errors.push(format!(
                "{label}.{name}: must be an integer with a minimum of at least 0"
            ));
        }
    }
    for name in names(&member["booleans"]) {
        if resolved(root, property(schema, &name))["type"] != "boolean" {
            errors.push(format!("{label}.{name}: must be boolean"));
        }
    }
}

/// A member object such as the request meta, reached through the envelope
/// property the contract names.
fn member_object(
    documents: &[Document],
    envelope: Located<'_>,
    member: &Value,
    label: &str,
    errors: &mut Vec<String>,
) -> Option<Value> {
    let root = &documents[envelope.0].value;
    let name = member["property"].as_str()?;
    let schema = resolved(root, property(envelope.1, name));
    exact_object(schema, label, member, errors);
    field_shapes(root, schema, label, member, errors);
    Some(schema.clone())
}

pub(super) fn schema_contract<'a>(
    documents: &'a [Document],
    contract: &Value,
    errors: &mut Vec<String>,
) -> BTreeMap<String, Located<'a>> {
    let mut schemas = BTreeMap::new();
    let protocol = contract["protocol"].as_str().unwrap_or_default();
    let envelopes = contract["envelopes"]
        .as_object()
        .cloned()
        .unwrap_or_else(Map::new);
    for (kind, member) in &envelopes {
        let Some(found) = schema_with_property_const(documents, "type", kind) else {
            errors.push(format!(
                "schema: no '{kind}' envelope with properties.type.const == '{kind}'"
            ));
            continue;
        };
        let label = format!("schema {kind} envelope");
        exact_object(found.1, &label, member, errors);
        field_shapes(
            &documents[found.0].value,
            found.1,
            &format!("schema {kind}"),
            member,
            errors,
        );
        schemas.insert(kind.clone(), found);
    }
    if let Some(&request) = schemas.get("request") {
        let meta_contract = &contract["requestMeta"];
        if let Some(meta) = member_object(
            documents,
            request,
            meta_contract,
            "schema request.meta",
            errors,
        ) {
            let constant = meta_contract["protocolConstant"]
                .as_str()
                .unwrap_or_default();
            if meta["properties"][constant]["const"].as_str() != Some(protocol) {
                errors.push(format!(
                    "schema request.meta.{constant}: const must be {protocol:?}"
                ));
            }
        }
    }
    if let Some(&error) = schemas.get("error") {
        member_object(
            documents,
            error,
            &contract["errorPayload"],
            "schema error.error",
            errors,
        );
    }
    let replay = &contract["replay"];
    let method = replay["method"].as_str().unwrap_or_default();
    if schema_with_property_const(documents, "method", method).is_none() {
        errors.push(format!(
            "schema: no request specialization with method const '{method}'"
        ));
    }
    let required = names(&replay["required"]);
    let optional = names(&replay["optional"]);
    let wanted = required.iter().chain(optional.iter()).collect::<Vec<_>>();
    match object_with_properties(documents, &wanted) {
        None => errors.push(format!("schema: no {method} params object with {wanted:?}")),
        Some((index, params)) => {
            exact_object(params, &format!("schema {method} params"), replay, errors);
            field_shapes(
                &documents[index].value,
                params,
                "schema replay",
                replay,
                errors,
            );
        }
    }
    schemas
}
